### PR TITLE
feat(backends): use PEP544 protocols for structural subtyping

### PR DESCRIPTION
**Pull request by lmilbaum**
The purpose of this change is to track [api](https://github.com/python-gitlab/python-gitlab/blob/main/docs/api-levels.rst) changes. For example: package versioning and breaking change announcement in case of protocol change.
This is MVP implementation to be used by #2435 
Haven't figured out yet how to implement unit tests for a protocol and its value.

### CODE REVIEW

File: .github/workflows/protocol.yml
Comment (nejch) [inline .github/workflows/protocol.yml:20]: Perhaps we can use one of our existing test suites if we write tests for this.

See also my comment on your other PR - if it maybe makes sense to do this at the backend level, and hide some of this from the client.

Comment (lmilbaum) [reply .github/workflows/protocol.yml:20]: I can't figure out how to implement a test which will identify the breaking change if a contributor would update the protocol.  The maintainer will have to manually watch for protocol change and be alerted. That is what this PR was about. Do you have ideas how to set it up?

File: gitlab/protocols/__init__.py
Comment (nejch) [inline gitlab/protocols/__init__.py:3]: Same as above: rather than having a separate filepath for this I think this belongs along with the backends in `gitlab/_backends`. Probably both in something like `gitlab/_backends/base.py` to keep it simple for now, also simplifies the imports :)

**Review by nejch — CHANGES_REQUESTED** [review]
Thanks @lmilbaum, I'm quite slow with the reviews here. Just a small note on keeping things simple for now.

File: gitlab/protocols/http_backend.py
Comment (JohnVillalovos) [inline gitlab/protocols/http_backend.py:2]: I personally (I think @nejch may have a different opinion) am not a fan of importing methods/classes.

I think just `import abc` is good and then easier when reading the code below to know where it comes from.

This is also recommended by Google's Python Style Guide, though I am not suggesting we adopt Google's Python Style Guide 🙂 And yes I realize that we are breaking this guideline in our current code 😥

https://google.github.io/styleguide/pyguide.html

2.2 Imports

Use import statements for packages and modules only, not for individual classes or functions.

File: gitlab/protocols/http_backend_response.py
Comment (JohnVillalovos) [inline gitlab/protocols/http_backend_response.py:2]: I personally (I think @nejch may have a different opinion) am not a fan of importing methods/classes.

I think just `import abc` is good and then easier when reading the code below to know where it comes from.

**Review by JohnVillalovos — COMMENTED** [review]
Thanks. Can the commit message have some details on the "why" of this PR?

**Comment by JohnVillalovos** [comment]
> Thanks. Can the commit message have some details on the "why" of this PR?

I still would like to see the description be put into the commit message.

At the moment all I see is:
```
$ git show --no-patch
commit 958d35d3c3d1755edf651cb15f5d1397ca571cc6 (HEAD -> refs/heads/protocol, refs/remotes/origin/protocol)
Author: Liora Milbaum <liora@lmb.co.il>
Date:   Mon Dec 26 06:39:58 2022 +0200

    feat: PEP 544 – Protocols: Structural subtyping (static duck typing)

```

File: gitlab/_backends/protocol.py
Comment (JohnVillalovos) [inline gitlab/_backends/protocol.py:16]: Has this idea been tested? I mean if we want to use `httpx` instead of `requests` how will this look?

I'm assuming that the idea is that we will be able to use `BackendResponse` and `Backend` in place of both `httpx` and `requests`. So wondering how that will look. Just wanting to make sure that this will work for that case.

Comment (lmilbaum) [reply gitlab/_backends/protocol.py:16]: This represent the existing interface which is `requests` only based. The plan is to change the interface as you described as soon as `httpx` would be introduced. Hoping that this answers your concern.

Comment (nejch) [reply gitlab/_backends/protocol.py:16]: Let's cross that bridge when we get there, we can still change the approach if it doesn't work - otherwise if we wanted to test with httpx already, we'd have a chunky PR that would also be hard to review :sweat_smile:

Comment (nejch) [reply gitlab/_backends/protocol.py:16]: But I think if the protocol defines all the methods and attributes we might be ok. It will be a problem where httpx and requests differ, for example `iter_content` vs `iter_bytes` etc. But we can maybe have little wrappers for that.

**Comment by nejch** [comment]
> > Thanks. Can the commit message have some details on the "why" of this PR?
> 
> I still would like to see the description be put into the commit message.
> 
> At the moment all I see is:

As an alternative we can squash-merge and it will link to the PR which has much more context than we would usually put in commit messages. WDYT @JohnVillalovos

**Comment by JohnVillalovos** [comment]
> As an alternative we can squash-merge and it will link to the PR which has much more context than we would usually put in commit messages. WDYT @JohnVillalovos

That is okay if it is in the commit message. I want the details in the Git log. Not only on GitHub. What if the project gets moved to GitLab in the future. That information can disappear if it isn't in the git repository.

Seems relatively easy to take the description above and put it in the commit message. But maybe I am missing something?

**Comment by nejch** [comment]
Sure, we can maybe copy it during the squash merge in the description field and it'll appear in the body. I've already sneakily edited the PR title a bit :smiley_cat: 

GitLab's importer is actually quite good these days so it would preserve the commit-to-MR associations, I personally like  that over long commit bodies as it includes the review context but that's subjective of course.

I'll approve for now, and see if there's anything else to cover :)

**Review by nejch — APPROVED** [review]

File: gitlab/_backends/protocol.py
Comment (nejch) [inline gitlab/_backends/protocol.py:11]: Actually, just one more comment. This won't work on 3.7 without a change in `setup.py`. But I'll open a follow-up for that.

File: gitlab/_backends/protocol.py
Comment (nejch) [reply gitlab/_backends/protocol.py:16]: Just to get a bit of confidence in the direction here, I created an httpx backend locally to take this for a spin. It passes almost all functional tests except a few things:
* httpx does not coerce query params to strings, which requests seems to do, so for example Enums have to explicitly provide the value
* `response_content` is backend-specific, and shouldn't live in utils
* streamed responses dont work out of the box, since it's handled differently
* multipart is also handled differently, so it doesn't work
* I completely skirted the typing issue, that will probably be a headache.
* we cannot run unit tests on both backends unless we do some wrapping of `responses`/`respx` libraries based on the backend.

Simple requests already work though. /cc @JohnVillalovos @lmilbaum

**Comment by nejch** [comment]
> > As an alternative we can squash-merge and it will link to the PR which has much more context than we would usually put in commit messages. WDYT @JohnVillalovos
> 
> That is okay if it is in the commit message. I want the details in the Git log. Not only on GitHub. What if the project gets moved to GitLab in the future. That information can disappear if it isn't in the git repository.
> 
> Seems relatively easy to take the description above and put it in the commit message. But maybe I am missing something?

I added the description in the body and reworded a bit during merge, https://github.com/python-gitlab/python-gitlab/commit/4afeaff0361a966254a7fbf0120e93583d460361.

**Pull request merged**
